packages/dds/sequence/src/test/intervalCollection.undoRedo.spec.ts

```diff
@@ -0,0 +1,165 @@
+
+const assertIntervals = (
+	sharedString: SharedString,
+	intervalCollection: IntervalCollection<SequenceInterval>,
+	expected: readonly { start: number; end: number }[],
+	validateOverlapping: boolean = true,
+) => {
+	const actual = Array.from(intervalCollection);
+	if (validateOverlapping && sharedString.getLength() > 0) {
+		const overlapping = intervalCollection.findOverlappingIntervals(
+			0,
+			sharedString.getLength() - 1,
+		);
+		assert.deepEqual(actual, overlapping, "Interval search returned inconsistent results");
+	}
+	assert.strictEqual(
+		actual.length,
+		expected.length,
+		`findOverlappingIntervals() must return the expected number of intervals`,
+	);
+
+	const actualPos = actual.map((interval) => {
+		assert(interval);
+		const start = sharedString.localReferencePositionToPosition(interval.start);
+		const end = sharedString.localReferencePositionToPosition(interval.end);
+		return { start, end };
+	});
+	assert.deepEqual(actualPos, expected, "intervals are not as expected");
+};
+
+function assertIntervalEquals(
+	string: SharedString,
+	interval: SequenceInterval | undefined,
+	endpoints: { start: number; end: number },
+): void {
+	assert(interval);
+	assert.equal(
+		string.localReferencePositionToPosition(interval.start),
+		endpoints.start,
+		"mismatched start",
+	);
+	assert.equal(
+		string.localReferencePositionToPosition(interval.end),
+		endpoints.end,
+		"mismatched end",
+	);
+}
+
+describe("Undo/redo for interval collection operations", () => {
+	let sharedString: SharedString;
+	let dataStoreRuntime1: MockFluidDataStoreRuntime;
+	// not sure if i need the factory if im only using one sharedstring
+	let containerRuntimeFactory: MockContainerRuntimeFactory;
+	let collection: IntervalCollection<SequenceInterval>;
+	let revertibles: MergeTreeDeltaRevertible[];
+
+	beforeEach(() => {
+		dataStoreRuntime1 = new MockFluidDataStoreRuntime({ clientId: "1" });
+		sharedString = new SharedString(
+			dataStoreRuntime1,
+			"shared-string-1",
+			SharedStringFactory.Attributes,
+		);
+
+		// unclear if this is needed
+		containerRuntimeFactory = new MockContainerRuntimeFactory();
+
+		// Connect the first SharedString.
+		dataStoreRuntime1.local = false;
+		const containerRuntime1 = containerRuntimeFactory.createContainerRuntime(dataStoreRuntime1);
+		const services1 = {
+			deltaConnection: containerRuntime1.createDeltaConnection(),
+			objectStorage: new MockStorage(),
+		};
+		sharedString.initializeLocal();
+		sharedString.connect(services1);
+		collection = sharedString.getIntervalCollection("test");
+		revertibles = [];
+	});
+
+	it("has an interval contained within the deleted range", () => {
+		sharedString.insertText(0, "hello world, this is me");
+
+		sharedString.on("sequenceDelta", (op) => {
+			appendToMergeTreeDeltaRevertibles(sharedString, op.deltaArgs, revertibles);
+		});
+
+		collection.add(6, 11, IntervalType.SlideOnRemove);
+
+		sharedString.removeRange(5, 11);
```

Comment:
Is the interval within the removed range if remove is exclusive of the end and interval add is inclusive?

Reply:
I was going off of the example from the ADO item, but figured it wouldn't hurt to change it when I shortened the string.

Reply:
> Is the interval within the removed range if remove is exclusive of the end and interval add is inclusive?

I spread some misinformation here in the past that I want to clarify. Though we add intervals in a way that makes their endpoints *look* inclusive (we will put endpoints on indices 6 and 11 here), if you consider how they behave over time from usage perspective, they act more like half-open: if the interval actually included both endpoints (so referred to "world,") and you executed `removeRange(10, 13)` (removing "d, "), your interval endpoint would now be on the "t". Clearly it's not desirable to have the interval "worlt" here (it should be "worl"), so it acts half-open.

There are a few places in the code which still have bits of this misinformation. I'm going to make a PR now to try and clarify them, probably with an example like the above as a remark.

With that said, I agree with Scarlett that the test case is less interesting than when both endpoints are within the range. It looks like you have that test case below (but mislabeled, accounting for some of the above).

Reply:
https://github.com/microsoft/FluidFramework/pull/15050

Reply:
something like this:
"has an interval contained within the deleted range" -> "has an interval with one endpoint contained within the deleted range"

"has an interval with the same range as the deleted text" -> "has an interval with both endpoints contained within the deleted range"

Reply:
Would one test be sufficient for having one interval endpoint in the deleted range then? (instead of having one each for interval start and end within the deleted range)